maro/cli/data_pipeline/vm_scheduling.py

```diff
@@ -0,0 +1,296 @@
+            vm_id_map = vm_id_map[vm_id_map.isin(vm_table['vmid'])]
+
+        return vm_id_map, vm_table
+
+    def _convert_cpu_readings_id(self, old_data_path: str, new_data_path: str, vm_id_map: pd.DataFrame):
+        """Convert vmid in each cpu readings file."""
+        with open(old_data_path, 'r') as f_in:
+            csv_reader = reader(f_in)
+            with open(new_data_path, 'w') as f_out:
+                csv_writer = writer(f_out)
+                csv_writer.writerow(['timestamp', 'vmid', 'maxcpu'])
+                for row in csv_reader:
+                    # [timestamp, vmid, mincpu, maxcpu, avgcpu]
+                    if row[1] in vm_id_map:
+                        new_row = [int(row[0]) // 300, vm_id_map.loc[row[1]], round(float(row[3]), 2)]
+                        csv_writer.writerow(new_row)
+
+    def _preprocess(self):
+        logger.info_green("Process vmtable data.")
+        # Process vmtable file.
+        vm_id_map, vm_table = self._process_vm_table(raw_vm_table_file=self._raw_vm_table_file)
+        with open(self._clean_file, mode="w", encoding="utf-8", newline="") as f:
+            vm_table.to_csv(f, index=False, header=True)
+
+        logger.info_green("Reading cpu data.")
+        # Process every cpu readings file.
+        for clean_cpu_readings_file_name in self._clean_cpu_readings_file_name_list:
+            raw_cpu_readings_file_name = clean_cpu_readings_file_name.split(".")[0] + "_raw.csv"
+            raw_cpu_readings_file = os.path.join(self._raw_folder, raw_cpu_readings_file_name)
+            clean_cpu_readings_file = os.path.join(self._clean_folder, clean_cpu_readings_file_name)
+            # Convert vmid.
+            logger.info_green(f"Process {clean_cpu_readings_file}.")
+            self._convert_cpu_readings_id(
+                old_data_path=raw_cpu_readings_file,
+                new_data_path=clean_cpu_readings_file,
+                vm_id_map=vm_id_map
+            )
+
+    def build(self):
+        super().build()
+        for clean_cpu_readings_file_name in self._clean_cpu_readings_file_name_list:
+            clean_cpu_readings_file = os.path.join(self._clean_folder, clean_cpu_readings_file_name)
+            if os.path.exists(clean_cpu_readings_file):
+                build_file_name = clean_cpu_readings_file_name.split(".")[0] + ".bin"
+                build_file = os.path.join(self._build_folder, build_file_name)
+                logger.info_green(f"Building binary data from {clean_cpu_readings_file} to {build_file}.")
+                cpu_meta_file = os.path.join(self._meta_folder, "cpu_readings.yml")
+                convert(meta=cpu_meta_file, file=[clean_cpu_readings_file], output=build_file)
+            else:
+                logger.warning_yellow(f"Not found cleaned data: {self._clean_file}.")
+
+
+class VmSchedulingTopology(DataTopology):
+    def __init__(self, topology: str, source: str, sample: int, seed: int, is_temp=False):
```

Comment:
the different topologies will have different sample rates right?

Reply:
Yes. It can be set at `meta/source_urls.yml`.